File: backend/services/identity-service/src/main/java/tn/bewerbi/identity/profile/ProfileCompleteness.java
package tn.bewerbi.identity.profile;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import tn.bewerbi.identity.domain.Profile;

/**
 * Source of truth for profile-completeness scoring, mirrored by the Flutter &
 * Web clients. Returns both the overall percent and the highest-value missing
 * field so the UI can prompt the user with the most impactful next action.
 */
public final class ProfileCompleteness {

    private ProfileCompleteness() {}

    public record MissingField(String key, int weight, String labelKey, String actionKey, String route) {}

    public record Result(int percent, Tier tier, List<MissingField> missing, MissingField nextAction) {}

    public enum Tier {
        STARTER, MOVER, ADVANCED, COMPLETE;
        public static Tier of(int percent) {
            if (percent >= 100) return COMPLETE;
            if (percent >= 75) return ADVANCED;
            if (percent >= 40) return MOVER;
            return STARTER;
        }
    }

    public static Result compute(Profile p) {
        var checks = List.of(
                new Check("firstName", 10, "profile.firstName", "profileAction.addFirstName", "/profile/edit", notBlank(p.getFirstName())),
                new Check("lastName", 10, "profile.lastName", "profileAction.addLastName", "/profile/edit", notBlank(p.getLastName())),
                new Check("phone", 8, "profile.phone", "profileAction.addPhone", "/profile/edit", notBlank(p.getPhone())),
                new Check("city", 6, "profile.city", "profileAction.addCity", "/profile/edit", notBlank(p.getCity())),
                new Check("country", 4, "profile.country", "profileAction.addCountry", "/profile/edit", notBlank(p.getCountry())),
                new Check("bio", 12, "profile.bio", "profileAction.addBio", "/profile/edit", notBlank(p.getBio())),
                new Check("photo", 10, "profile.photo", "profileAction.addPhoto", "/profile/edit", notBlank(p.getPhotoUrl())),
                new Check("desiredProfession", 10, "onboarding.step1Title", "profileAction.addProfession", "/onboarding", notBlank(p.getDesiredProfession())),
                new Check("germanLevel", 15, "profile.level", "profileAction.addGerman", "/onboarding", p.getGermanLevel() != null),
                new Check("recognitionStatus", 5, "anerkennung.title", "profileAction.addRecognition", "/anerkennung", p.getRecognitionStatus() != null),
                new Check("skills", 10, "onboarding.step4Title", "profileAction.addSkills", "/onboarding", p.getSkills() != null && !p.getSkills().isEmpty())
        );

        int percent = 0;
        var missing = new ArrayList<MissingField>();
        for (Check c : checks) {
            if (c.has) {
                percent += c.weight;
            } else {
                missing.add(new MissingField(c.key, c.weight, c.labelKey, c.actionKey, c.route));
            }
        }
        percent = Math.min(100, percent);
        MissingField next = missing.stream()
                .max(Comparator.comparingInt(MissingField::weight))
                .orElse(null);
        return new Result(percent, Tier.of(percent), missing, next);
    }

    private static boolean notBlank(String s) { return s != null && !s.isBlank(); }

    private record Check(String key, int weight, String labelKey, String actionKey, String route, boolean has) {}
}

File: mobile/src/lib/profileCompleteness.ts
/**
 * Profile-completeness field weights — single source of truth for the web client.
 * Mirror of the Java `tn.bewerbi.profile.ProfileCompleteness` class.
 */
export interface ProfileForCompleteness {
  first_name?: string | null;
  last_name?: string | null;
  phone?: string | null;
  city?: string | null;
  country?: string | null;
  bio?: string | null;
  photo_url?: string | null;
  desired_profession?: string | null;
  german_level?: string | null;
  recognition_status?: string | null;
  skills?: string[];
}

interface Field {
  key: string;
  weight: number;
  labelKey: string;
  actionKey: string;
  route: string;
  check: (p: ProfileForCompleteness) => boolean;
}

const FIELDS: Field[] = [
  { key: "firstName", weight: 10, labelKey: "auth.firstName",
    actionKey: "profileAction.addFirstName", route: "/(applicant)/(profile)/edit",
    check: (p) => !!p.first_name?.trim() },
  { key: "lastName", weight: 10, labelKey: "auth.lastName",
    actionKey: "profileAction.addLastName", route: "/(applicant)/(profile)/edit",
    check: (p) => !!p.last_name?.trim() },
  { key: "phone", weight: 8, labelKey: "profile.phone",
    actionKey: "profileAction.addPhone", route: "/(applicant)/(profile)/edit",
    check: (p) => !!p.phone?.trim() },
  { key: "city", weight: 6, labelKey: "profile.city",
    actionKey: "profileAction.addCity", route: "/(applicant)/(profile)/edit",
    check: (p) => !!p.city?.trim() },
  { key: "country", weight: 4, labelKey: "profile.country",
    actionKey: "profileAction.addCountry", route: "/(applicant)/(profile)/edit",
    check: (p) => !!p.country?.trim() },
  { key: "bio", weight: 12, labelKey: "profile.bio",
    actionKey: "profileAction.addBio", route: "/(applicant)/(profile)/edit",
    check: (p) => !!p.bio?.trim() },
  { key: "photo", weight: 10, labelKey: "profile.photo",
    actionKey: "profileAction.addPhoto", route: "/(applicant)/(profile)/edit",
    check: (p) => !!p.photo_url?.trim() },
  { key: "desiredProfession", weight: 10, labelKey: "onboarding.step1Title",
    actionKey: "profileAction.addProfession", route: "/(applicant)/onboarding",
    check: (p) => !!p.desired_profession?.trim() },
  { key: "germanLevel", weight: 15, labelKey: "profile.level",
    actionKey: "profileAction.addGerman", route: "/(applicant)/onboarding",
    check: (p) => !!p.german_level },
  { key: "recognitionStatus", weight: 5, labelKey: "anerkennung.title",
    actionKey: "profileAction.addRecognition", route: "/(applicant)/anerkennung",
    check: (p) => !!p.recognition_status },
  { key: "skills", weight: 10, labelKey: "onboarding.step4Title",
    actionKey: "profileAction.addSkills", route: "/(applicant)/onboarding",
    check: (p) => (p.skills ?? []).length > 0 },
];

export interface CompletenessResult {
  percent: number;
  tier: CompletenessTier;
  missing: MissingField[];
  nextAction: MissingField | null;
}

export interface MissingField {
  key: string;
  weight: number;
  labelKey: string;
  actionKey: string;
  route: string;
}

export type CompletenessTier = "starter" | "mover" | "advanced" | "complete";

export function compute(profile: ProfileForCompleteness | null): CompletenessResult {
  if (!profile) return { percent: 0, tier: "starter", missing: [], nextAction: null };

  let percent = 0;
  const missing: MissingField[] = [];
  for (const f of FIELDS) {
    if (f.check(profile)) {
      percent += f.weight;
    } else {
      missing.push({
        key: f.key,
        weight: f.weight,
        labelKey: f.labelKey,
        actionKey: f.actionKey,
        route: f.route,
      });
    }
  }
  percent = Math.min(100, percent);

  // Next action: heaviest missing field (highest impact)
  const nextAction = missing.length
    ? [...missing].sort((a, b) => b.weight - a.weight)[0]
    : null;

  return { percent, tier: tierFor(percent), missing, nextAction };
}

export function tierFor(percent: number): CompletenessTier {
  if (percent >= 100) return "complete";
  if (percent >= 75) return "advanced";
  if (percent >= 40) return "mover";
  return "starter";
}

export const TIER_META: Record<
  CompletenessTier,
  { labelKey: string; color: string; bgClass: string; emoji: string }
> = {
  starter: { labelKey: "tier.starter", color: "#6b7280", bgClass: "bg-gray-100", emoji: "🌱" },
  mover: { labelKey: "tier.mover", color: "#2563EB", bgClass: "bg-primary-50", emoji: "🚀" },
  advanced: { labelKey: "tier.advanced", color: "#7c3aed", bgClass: "bg-purple-50", emoji: "⭐" },
  complete: { labelKey: "tier.complete", color: "#16a34a", bgClass: "bg-success-100", emoji: "🏆" },
};
